server/src/server.rs

```diff
@@ -146,7 +152,22 @@ impl Server {
             .check_n(NonZeroU32::new(packet.payload.len().try_into().unwrap_or(u32::MAX)).unwrap())
         {
             log::trace!("Rate limiting: {:?}", e);
-            // TODO: Send `PauseForwarding`
+            match e {
+                NegativeMultiDecision::InsufficientCapacity(cells) => {
+                    // the query was invalid as the rate limite parameters can never accomodate the
+                    // number of cells queried for.
+                    log::warn!("Rate limited packet dropped. Exceeds limit. size: {}, from: {}", cells, &from);
+                },
+                NegativeMultiDecision::BatchNonConforming(cells, retry_at) => {
+                    log::debug!("Rate limited packet. size: {}, retry_at: {}", cells, retry_at);
+                    let mut server = self.state.write().await;
+                    server.resume_forwarding.insert((retry_at.earliest_possible(), session_id.clone(), from.clone()));
```

Comment:
```suggestion
                    {
                        let mut server = self.state.write().await;
                        server.resume_forwarding.insert((retry_at.earliest_possible(), session_id.clone(), from.clone()));
                    }
```
Drop the write lock asap